### ISSUE TITLE
feat: Async DataLoader with prefetching for GPU/CPU pipeline overlap

### ISSUE THREAD

**Issue by Samyssmile** [Samyssmile/edux#195]
## Summary

The current `DataLoader` is fully synchronous — each batch is assembled **during** the training loop, blocking the entire pipeline. While the GPU/CPU processes the current batch (forward, backward, optimizer step), the DataLoader sits idle. This creates a severe throughput bottleneck, especially for ImageNet-scale training with disk-resident images.

**Estimated impact**: 2–5× training throughput improvement depending on dataset I/O characteristics.

## Current Architecture (Problem)

```
Thread: [fetch batch 1][=== train batch 1 ===][fetch batch 2][=== train batch 2 ===]
GPU:    [    idle     ][======= busy ========][    idle     ][======= busy ========]
```

### Current flow in `DataLoader.java`:
```java
// BatchIterator.next() — fully blocking
public Batch next() {
    var batchData = new float[currentBatchSize * dataFlatSize];    // fresh alloc per batch
    var batchLabels = new float[currentBatchSize * labelFlatSize]; // fresh alloc per batch
    
    for (int i = 0; i < currentBatchSize; i++) {
        int idx = indices[position++];
        var sampleData = dataset.getData(idx);     // ← BLOCKING I/O
        var sampleLabel = dataset.getLabel(idx);    // ← BLOCKING I/O
        System.arraycopy(...);
    }
    return new Batch(data, labels);
}
```

### Current flow in `Trainer.java`:
```java
for (var batch : trainLoader) {        // ← blocks here until batch is ready
    var predictions = model.forward(batch.data(), true);   // GPU busy
    var lossResult = loss.compute(predictions, batch.labels());
    model.backward(lossResult.gradient());
    optimizer.step(layers);
    model.clearCaches();
    // → loop back: blocks again on next batch fetch
}
```

### Key problems:

1. **No pipelining**: GPU computation and data loading cannot overlap
2. **Sequential sample fetching**: Samples loaded one-by-one in a for-loop
3. **No buffer reuse**: Fresh `float[]` arrays allocated every batch (GC pressure)
4. **No parallel decode**: Image decoding (ImageIO.read, resize, transform) is single-threaded
5. **First-sample-only prefetch**: Only 1 sample prefetched at construction for shape inference

## Proposed Architecture

```
Prefetch Thread(s): [fetch B2][fetch B3][fetch B4][ ...waiting... ][fetch B5]...
Training Thread:    [=== train B1 ===][=== train B2 ===][=== train B3 ===]...
GPU:                [===== busy =====][===== busy =====][===== busy =====]...
```

### Design Goals

1. **Zero-wait batch delivery**: Next batch ready before current training step finishes
2. **Bounded memory**: Fixed-size prefetch queue (configurable depth, default 2–3)
3. **Backward-compatible API**: `DataLoader` still implements `Iterable<Batch>`
4. **JDK 25 idiomatic**: Leverage Virtual Threads (Project Loom) for lightweight prefetching
5. **GPU-aware**: Optional pre-transfer to GPU memory during prefetch (future)

### Proposed API

```java
// Option 1: Builder pattern (recommended)
var loader = DataLoader.builder(dataset, batchSize)
    .shuffle(true)
    .prefetch(3)              // prefetch queue depth (0 = sync, default)
    .numWorkers(4)            // parallel sample decode workers
    .build();

// Option 2: Backward-compatible constructor overload
var loader = new DataLoader(dataset, batchSize, shuffle, prefetchCount);

// Usage unchanged:
for (var batch : loader) {
    model.forward(batch.data(), true);
    // ...
}

// AutoCloseable for clean shutdown:
try (var loader = DataLoader.builder(...).prefetch(3).build()) {
    trainer.train(loader, testLoader, config);
}
```

### Internal Architecture

```
┌─────────────────────────────────────────────────────────┐
│                    DataLoader                           │
│                                                         │
│  ┌──────────────┐     ┌──────────────────────────────┐  │
│  │ Prefetch Pool │────▶│ BlockingQueue<Batch>(depth)  │  │
│  │ (Virtual      │     │                              │  │
│  │  Threads)     │     │  [Batch 3] [Batch 2] [ready] │──┼──▶ iterator.next()
│  │               │     │                              │  │     (non-blocking if
│  │ Worker 1 ─┐   │     └──────────────────────────────┘  │      queue non-empty)
│  │ Worker 2 ─┤   │                                       │
│  │ Worker N ─┘   │     ┌──────────────────────────────┐  │
│  │               │     │ BufferPool<float[]>           │  │
│  │ Each worker:  │     │ Reusable batch arrays         │  │
│  │ 1. getData()  │     │ (avoids GC pressure)          │  │
│  │ 2. getLabel() │     └──────────────────────────────┘  │
│  │ 3. assemble   │                                       │
│  │ 4. put(queue) │                                       │
│  └──────────────┘                                        │
└─────────────────────────────────────────────────────────┘
```

### Phase 1: Core Async Prefetching
- [ ] Add `prefetchCount` parameter to `DataLoader` (default `0` = current sync behavior)
- [ ] Implement `PrefetchIterator` using `LinkedBlockingQueue<Batch>(prefetchCount)`
- [ ] Use Virtual Threads (`Thread.ofVirtual()`) for prefetch workers
- [ ] Implement `AutoCloseable` on `DataLoader` for clean executor shutdown
- [ ] Poison-pill pattern for iterator exhaustion signaling
- [ ] Proper exception propagation from prefetch thread to consumer

### Phase 2: Parallel Sample Decoding
- [ ] Add `numWorkers` parameter for parallel `dataset.getData()` calls within a batch
- [ ] Use `StructuredTaskScope` (JDK 25 preview) for scoped parallelism per batch
- [ ] Parallel image decode is especially impactful for `ImageFolderDataset` (ImageIO.read + resize)

### Phase 3: Buffer Pooling (Optional)
- [ ] Implement `BatchBufferPool` that recycles `float[]` arrays
- [ ] `Batch` gets a `release()` method to return buffers to pool
- [ ] Trainer calls `batch.release()` after `model.clearCaches()` — or auto-release via `try-with-resources`

### Thread Safety
- `Dataset.getData()` must be thread-safe for parallel decoding
- `ImageDataset` uses lazy caching (`dataCache[index]`) — needs synchronization or `AtomicReferenceArray`
- `AugmentedDataset` uses `ThreadLocalRandom` — already thread-safe
- Shuffle indices must be generated before prefetch starts (current design already does this)

### Shutdown & Resource Cleanup
```java
public void close() {
    prefetchExecutor.shutdownNow();
    prefetchQueue.clear();
    // Interrupt blocked prefetch threads
}
```

### Error Handling
- Prefetch thread wraps exceptions in a sentinel `Batch` or `CompletableFuture`
- Consumer re-throws on `next()` if prefetch encountered an error
- Timeout on `queue.poll()` to detect stuck prefetch threads

### Backward Compatibility
- Default `prefetchCount = 0` preserves current synchronous behavior
- Existing `new DataLoader(dataset, batchSize, shuffle)` constructor unchanged
- `Trainer.java` needs **zero changes** — `Iterable<Batch>` contract preserved

## Performance Expectations

| Scenario | Current | With Prefetch (depth=2) | Speedup |
|----------|---------|------------------------|---------|
| MNIST (cached, CPU) | ~baseline | ~1.1× | Minimal (compute-bound) |
| ImageNet 64×64 (disk, GPU) | ~baseline | ~2-3× | I/O overlaps compute |
| ImageNet 224×224 (disk, GPU) | ~baseline | ~3-5× | Heavy I/O + decode |
| ImageNet + Augmentation | ~baseline | ~2-4× | Augment in prefetch |

## Acceptance Criteria

- [ ] `DataLoader` with `prefetch > 0` delivers batches without blocking when queue is non-empty
- [ ] Training throughput measurably improves on `ImageFolderDataset` with disk I/O
- [ ] `prefetch = 0` behavior is identical to current implementation (no regression)
- [ ] Clean shutdown: no leaked threads, no hanging on `close()`
- [ ] Exception from `Dataset.getData()` propagates correctly to training loop
- [ ] Unit test: verify prefetch queue fills ahead of consumption
- [ ] Unit test: verify correct batch ordering with shuffle
- [ ] Integration test: train a small model with async DataLoader and verify convergence

## References

- PyTorch DataLoader: `num_workers`, `prefetch_factor` parameters
- JDK 25 Virtual Threads: [JEP 444](https://openjdk.org/jeps/444)
- JDK 25 Structured Concurrency: [JEP 480](https://openjdk.org/jeps/480)

**Comment by Samyssmile** [comment]
Implemented and merged: async DataLoader with virtual-thread prefetching, StructuredTaskScope parallel sample decoding, builder API, SIMD Transforms, RandomResizedCrop, and full test coverage.
